src/gtk/dc.cpp

```diff
@@ -237,6 +240,76 @@ void* wxGTKCairoDCImpl::GetCairoContext() const
     return cr;
 }
 
+bool wxGTKCairoDCImpl::TryGetWindowSize(wxSize& size, wxSize& sizeMM) const
+{
+    if ( !m_window )
+        return false;
+
+    GdkWindow* const w = gtk_widget_get_window(m_window->GetHandle());
+    if ( !w )
+        return false;
+
+#if GTK_CHECK_VERSION(3, 22, 0)
+    if ( wx_is_at_least_gtk3(22) )
+    {
+        GdkDisplay* const d = gdk_window_get_display(w);
+        if ( !d )
+            return false;
+
+        GdkMonitor* const m = gdk_display_get_monitor_at_window(d, w);
+        if ( !m )
+            return false;
+
+        GdkRectangle rect;
+        gdk_monitor_get_geometry(m, &rect);
+
+        // The returned geometry is in the application pixels, convert them to
+        // the real device pixels to obtain the real PPI.
+        const int scale = gdk_monitor_get_scale_factor(m);
+        size.x = rect.width * scale;
+        size.y = rect.height * scale;
+
+        sizeMM.x = gdk_monitor_get_width_mm(m);
+        sizeMM.y = gdk_monitor_get_height_mm(m);
+
+        return true;
+    }
+#endif // GTK+ >= 3.22
+
+    // For GTK+4 the code above is always used, but for GTK+ 3 we fall back to
+    // it when run-time GTK+ version is < 3.22 or when compiled with < 3.22 in
+    // the first place.
+#ifndef __WXGTK4__
+    wxGCC_WARNING_SUPPRESS(deprecated-declarations)
+    GdkScreen* const s = gdk_window_get_screen(w);
+    if ( !s )
+        return false;
+
+    size.x = gdk_screen_get_width(s);
+    size.y = gdk_screen_get_height(s);
+    wxGCC_WARNING_RESTORE()
+#endif // GTK+ < 4
+
+    return true;
+}
+
+wxSize wxGTKCairoDCImpl::GetPPI() const
+{
+    wxSize size, sizeMM;
+    if ( TryGetWindowSize(size, sizeMM) )
+    {
+        return wxSize(wxRound((inches2mm * size.x) / sizeMM.x),
+                      wxRound((inches2mm * size.y) / sizeMM.y));
+
+    }
+    else
+    {
+        // For a non-window-based DC the concept of PPI doesn't make much sense
+        // anyhow, so just return the hardcoded value used by the base class.
+        return wxGCDCImpl::GetPPI();
```

Comment:
Sorry, I hadn't thought at all about who the author of the code was and, of course, I didn't mean to piss you off and I would have never raised this issue or touched this code if I thought it would result in it.

I admit to having been frustrated when writing this commit message because wxGTK DC classes are a mess (due to different headers and base classes used for the same class in GTK+ 2/3) and it took me much longer than it ought to have to understand how this code was structured and I assure you that not seeing explicit base class names contributed to it (it wasn't, of course, the only, or maybe even the major, factor, but it definitely didn't help).

I admit that I didn't think about the idea of using `base_type` to ensure that the base class method is not skipped, which may well have merit, but in my defence, it could be explained by the fact that it doesn't apply at all for the ctor initializer list (where using non-direct base class would be a compilation error).

Anyhow, sorry once again for not formulating this commit message in a more neutral way and the intention, if it counts, definitely wasn't to criticize anybody in particular, lest of all you. I'd still like to avoid using `base_type` in initializer lists to keep things simpler, if possible, but if you really want to have it, let me know and I'll revert this change. Also, using `BaseType` would be at least more consistent with wx naming conventions and not stick out that much.